guidelines/design-guidelines.md

```diff
@@ -31,4 +46,148 @@ Example:
 var transportConfig = endpointConfig.UseTransport<MsmqTransport>();
 
 transportConfig.MsmqLabelGenerator(context => return $"{context.Headers['NServiceBus.EnclosedMessageTypes']}");
-```
\ No newline at end of file
+```
+
+## Startable and stoppable components
+
+```
+class Component
+{
+  Task Start();
+  Task Stop();
+}
+```
+
+* All components externally implemented but managed by NServiceBus (pipelines, satellites, etc) should attempt to stop gracefully. If they they are unable to do so, they should block the thread.
+* NServiceBus will **not** timebox any start or stop operations. This allows proper debugging and analysis of misbehaving components.
+* Stop operations should be performed concurrently if possible/sensible to allow all components to initiate a graceful stop.
+* `CancellationToken` and `CancellationTokenSource` are an implementation detail of the component that needs to stop and are not managed by NServiceBus.
+
+Translating these principles into code, here is how NServiceBus handles startable and stoppable components:
+
+```
+var components = new List<Component>();
+
+// Start
+foreach(cmp in components)
+{
+  await cmp.Start().ConfigureAwait(false);
+}
+
+// Stop
+var stopTasks = components.Select(cmp => cmp.Stop());
+await Task.WhenAll(stopTasks);
+```
+
+
+## Namespace rules
+
+### Principles
+
+* Namespaces are used to guide external users and to uniquely identify types.
+* Folders are used to structure code for our internal purposes and don't necessarily align with the namespaces.
+* Our public API has two main consumers: developers building business solutions and developers extending NServiceBus. 
+
+### NServiceBus rules
+
+* For public types relevant to business developers, we use the `NServiceBus` namespace to make them more discoverable.
+* For internal types, we also use the `NServiceBus` namespace. This allows unique identification of those types in logs and stack traces.
+* For public types designed for extensibility, we use `NServiceBus.{ExtensibilityPoint}`. This hides types irrelevant to business developers while making them discoverable for developers extending NServiceBus. 
+   - Examples of currently used extensibility namespaces:
+      * `NServiceBus.Transport`
+      * `NServiceBus.Persistence`
+      * `NServiceBus.Serialization`
+      * `NServiceBus.Logging`
+      
+### Downstream components rules
+
+* For public types relevant to business developers, we use the `NServiceBus` namespace to make them more discoverable and to avoid extra using statements
+* For internal types, we use the root [`{Component}`](#Component naming rules) namespace. This allows for unique identification of those types in logs and stack traces.
```

Comment:
I don't think this anchor works (click "View" for this file). I think it has to be `#component-naming-rules`.

E.g. this link works: https://github.com/Particular/NServiceBus/blob/migrate-rules-from-platdev/guidelines/design-guidelines.md#component-naming-rules